### DISCUSSION TITLE
Inverse resource

### DISCUSSION THREAD

**Discussion by ftzi** [i18next/i18next#1575] [Q&A]
Is there any in-built way to create resources as

```js
{
  welcome: {
    en: 'Welcome, SrBrahma!',
    pt: 'Bem-vindo, SrBrahma!'
  }
}
```
?

I think this way is better for me, as I don't have too many texts and I don't want to miss any text from translating, and also, faster and simpler to write and to read.

For now, I will create a function to convert this format to the default

```js
{
  en: {
    translation: {
      welcome: 'Welcome, SrBrahma!',
    }
  },
  pt: {
    translation: {
      welcome: 'Bem-vindo, SrBrahma!',
    }
  }
}
```

**Comment by ftzi** [comment]
Here is my solution:

```ts
export type Languages = 'en' | 'pt';

type MyResource = {
  [textId: string]: {
    [language in Languages]: string
  }
}

function convertResource(myResource: MyResource): Resource {
  const converted: Record<string, any> = {}; // Resource wasn't working here
  for (const [textId, translations] of Object.entries(myResource))
    for (const [language, translation] of Object.entries(translations))
      converted[language].translation[textId] = translation;
  return converted;
}
```

**Reply by ftzi** [reply]
* Note that I don't really have a performance objective with those solutions, as I don't have too much written in my App (maybe ~100 translatable texts). Improving DX worths way more to me (and performance really shouldn't change at all).


Improving further the answer above, I also added a way to get the textId, so you won't input strings when using `t()`, so it's safer and better, as it uses Intellisense.

```ts
function createMyResource<T extends MyResource>(myResource: T): T {
  return myResource;
}

const myResources = createMyResource({
  welcome: {
    // en ,pt: ...
  }
})

function getTexts<T extends MyResource>(myResource: T): {[K in keyof T]: K} {
  const texts: Record<string, string> = {};
  for (const textId of Object.keys(myResource))
    texts[textId] = textId;
  return texts as {[K in keyof T]: K};
}

export const Texts = getTexts(myResources);
// Will have the result of Texts={welcome: 'welcome', etc1: 'etc1', etc2: 'etc2', ... }

// Usage
t(Texts.welcome)
```

Maybe I can even add nesting later. But it's good enough for now, to release the international version asap.

**Reply by ftzi** [reply]
Also, it works if you, in VsCode, press F2 to Rename Symbol (rename the "variable" everywhere it's used, in its proper context).

So you can press F2 in `welcome`, change to `hello`, and everywhere that was using `t(Texts.welcome)`, will change (safely) to `t(Texts.hello)`

**Reply by ftzi** [reply]
A Proxy could even be used, so instead of using `t(Texts.welcome)`, just `Texts.welcome` would do it without a performance impact (would translate the call, wouldn't create a function for each translation).

However, for react-i18n (which I uses), it would be required to include it in `useTranslation();` hook. I won't change this lib (maybe for now), but looks a good idea, for a better DX.

**Comment by jamuhl** [comment]
Does not make a lot of sense -> loading in backends is defined as lng->ns

Typescript gives already type-safety for keys during development

**Reply by ftzi** [reply]
Defined strings are type safe, however, they don't allow symbol renaming.

My i18n is client only, as is is an app
